mempool/src/core_mempool/ttl_cache.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    time::{Duration, SystemTime},
+};
+
+struct ValueInfo<V> {
+    value: V,
+    ttl: SystemTime,
+}
+
+pub struct TtlCache<K, V> {
+    capacity: usize,
+    default_timeout: Duration,
+    data: HashMap<K, ValueInfo<V>>,
+    ttl_index: BTreeMap<SystemTime, K>,
+}
+
+impl<K, V> TtlCache<K, V>
+where
+    K: std::cmp::Eq + std::hash::Hash + std::clone::Clone,
+{
+    pub fn new(capacity: usize, default_timeout: Duration) -> Self {
+        Self {
+            capacity,
+            default_timeout,
+            data: HashMap::new(),
+            ttl_index: BTreeMap::new(),
+        }
+    }
+
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.data.get(key).map(|v| &v.value)
+    }
+
+    pub fn insert(&mut self, key: K, value: V) {
+        self.gc(SystemTime::now());
+
+        // remove old entry if it exists
+        if let Some(info) = self.data.get(&key) {
+            self.ttl_index.remove(&info.ttl);
+        }
+
+        // insert new one
+        if let Some(expiration_time) = SystemTime::now().checked_add(self.default_timeout) {
+            self.ttl_index.insert(expiration_time.clone(), key.clone());
+            let value_info = ValueInfo {
+                value,
+                ttl: expiration_time,
+            };
+            self.data.insert(key, value_info);
+        }
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        self.gc(SystemTime::now());
+        match self.data.remove(&key) {
+            Some(info) => {
+                self.ttl_index.remove(&info.ttl);
+                Some(info.value)
+            }
+            None => None,
+        }
+    }
+
+    pub fn gc(&mut self, gc_time: SystemTime) {
```

Comment:
Could this be an expensive call over time? Might be worth only calling gc logarithmically either based upon size and/or time.

Reply:
switched it to be called directly from periodic gc task in shared mempoool